paddle/operators/recv_op.cc

```diff
@@ -61,29 +61,74 @@ class RecvOp : public framework::OperatorBase {
     server_thread_->join();
   }
 
+  std::string GetGradVarNameForTrainer(const std::string &varname) const {
+    if (grads_counter_.find(varname) == grads_counter_.end()) {
+      grads_counter_[varname] = 0;
+    }
+    char ret[256];
+    snprintf(ret, sizeof(ret), "%s.trainer_%d", varname.c_str(),
+             grads_counter_[varname]++);
+    return std::string(ret);
+  }
+
   void Run(const framework::Scope &scope,
            const platform::DeviceContext &dev_ctx) const override {
-    // blocking get one var from client.
-    const framework::LoDTensor &t = rpc_service_->Get();
+    // FIXME(typhoonzero): no new scopes for every run.
     framework::Scope &recv_scope = scope.NewScope();
-    // set graph input var
-    auto *var = recv_scope.Var(Input("RX"));
-    auto *tensor = var->GetMutable<framework::LoDTensor>();
-    // FIXME(typhoonzero): do not copy
-    framework::CopyFrom(t, dev_ctx.GetPlace(), dev_ctx, tensor);
-
-    std::string program_str = Attr<std::string>("OptimizeProgram");
-    framework::ProgramDesc program_desc;
-    program_desc.ParseFromString(program_str);
-    framework::ProgramDescBind program(program_desc);
-    framework::Executor executor(dev_ctx);
-    // Run sub graph to get optimized tensor
-    executor.Run(program, &recv_scope, 0, /*global_block*/
-                 false /*create_local_scope*/);
-
-    auto *out_var = recv_scope.FindVar("Out");
-    // push back
-    rpc_service_->Push(out_var->Get<framework::LoDTensor>());
+    rpc_service_->SetScope(&recv_scope);
+    auto param_list = Attr<std::vector<std::string>>("ParamList");
+    auto grad_list = Attr<std::vector<std::string>>("GradList");
```

Comment:
I think the `grad_list` and `param_list` are used to ensure pserver received a pair of parameter vs. gradient, and maybe we could do this in an Op which named `SyncWait`(or other name): checking the parameters vs. gradients in the blocking queue until received all tensors.

Reply:
I think if we use one send-recv op for one var, then we don't need to maintain two lists. Another benefit is here: https://github.com/PaddlePaddle/Paddle/pull/6297#discussion_r158121595

Reply:
Sure, but I think we also need a global Blocking Queue to store the variant so that recv op will fetch variant from the queue.

Reply:
@helinwang Indeed, `param_list` is not used in fact. The reason of we need to know the variable names is that we need to merge variables from multiple workers. The variable names from multiple workers are the same, so we need to rename them by adding numberic suffixes: `xxx@GRAD.trainer_0` to `xxx@GRAD.trainer_n` then the `sum_op` will use them as input.

The rename operation must happen before the sub program runs, and only `recv_op` can know when the variable arrives. I can't find a better way right now. But still, can make the `recv_op` more general by change the names.

Reply:
@Yancey1989 `param_list` is in fact not used for now, will remove.

The idea of using a queue before sending variables is great that we can send the variable parallelly right after the variable is ready to gain some performance.

Reply:
@typhoonzero I think we can merge this PR soon, this is great work we can merge in to the develop branch to start iterate.
In the same time, in my mind send-recv op is lower level than merging variables, it should be used by the merging variable logic, rather than controlling the merging variable logic.
You mentioned we need to have more "temp" variables, maybe they should be created by the transpiler, and send-recv op does not need to know it. For example, I saw `snprintf(ret, sizeof(ret), "%s.trainer_%d", varname.c_str()` in `recv_op.cc`, maybe it should be in the transpiler.

Reply:
Agree!